File: current_session/python/987.py
"""
https://leetcode.com/problems/vertical-order-traversal-of-a-binary-tree/
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import collections
class Solution(object):
    def __init__(self):
        self.nd = collections.defaultdict(list)

    def verticalTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        # naive approach:
        # step 1. give all nodes x value O(n)
        # step 2. put all nodes into a dictionary, key = x-value, value = node value
        # step 3. get the answer
        self.fillNd(root)
        xs = self.nd.keys()
        xs.sort()
        ans = []
        for x in xs:
            self.nd[x].sort()
            ans += [a[1] for a in self.nd[x]],
        return ans


    def fillNd(self, node, x=0, y=0):
        if not node: return
        self.nd[x].append((y,node.val))
        self.fillNd(node.left,x-1,y+1)
        self.fillNd(node.right,x+1,y+1)

File: current_session/c++/987.cpp
#include <iostream>
#include <vector>
#include <map>
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
public:
  vector<vector<int>> verticalTraversal(TreeNode* root) {
    // simply create a map and return sorted value
    helper(root, 0, 0);
    vector<vector<int>> res;
    for (int i=minPos; i<=maxPos; ++i) {
      vector<int> tmp;          // create tmp vector to store all nodes in pos i
      for (auto k: nodeMap[i]) {
        sort(k.second.begin(), k.second.end());     // sort nodes in same (depth/pos)
        for (auto x: k.second) {          // add nodes in different depth
          tmp.push_back(x);
        }
      }
      res.push_back(tmp);
    }
    return res;
  }

private:
  map<int, map<int, vector<int>> > nodeMap;        // key: pos, value: map- k:depth value:node values
  int minPos=0, maxPos=0;
  void helper(TreeNode* node, int depth, int pos) {
    // create a stack for each layer
    minPos = min(minPos, pos);
    maxPos = max(maxPos, pos);
    // nodeMap[pos].push_back(node->val);
    nodeMap[pos][depth].push_back(node->val);
    stack<TreeNode*> st;
    if (node->left != nullptr) helper(node->left, depth+1, pos-1);
    if (node->right != nullptr) helper(node->right, depth+1, pos+1);
    return;
  }
};